components/support/sql/src/conn_ext.rs

```diff
@@ -223,6 +223,15 @@ pub trait ConnExt {
     fn unchecked_transaction_imm(&self) -> SqlResult<UncheckedTransaction<'_>> {
         UncheckedTransaction::new(self.conn(), TransactionBehavior::Immediate)
     }
+
+    /// Get the approximate DB size in bytes
+    fn get_db_size(&self) -> Result<u32, rusqlite::Error> {
+        let page_count: u32 = self.query_one("SELECT * from pragma_page_count()")?;
+        let page_size: u32 = self.query_one("SELECT * from pragma_page_size()")?;
+        let freelist_count: u32 = self.query_one("SELECT * from pragma_freelist_count()")?;
+
+        Ok((page_count - freelist_count) * page_size)
+    }
```

Comment:
for interest, have you sanity checked how closely this matches the actual size on disk? I guess I'm really wondering if we should add more caveats to the docs.

Also wondering about the freelist - that still takes space on the disk, right? If this process just ends up creating a larger freelist without actually freeing up disk space, I wonder how useful it turns out being? Or do we assume the vacuum cleans that up?

Reply:
Astonishingly well.  I tested the moz-central test data in `toolkit/components/passwordmgr/test/unit/data/` and `toolkit/components/places/tests/migration/`.  The result of that calculation exactly matched the output of `du -b`, for all files except the corrupt one.

I don't think those files had any pages in the free list, so I ran this test to check that:
  - Created a database file, inserted a bunch of junk data, then deleted half of it
  - Checked the output of `du -b` against the query calculation.  The query calculation was almost exactly half of `du -b`
  - Ran vaccum and checked again.  This time the query calculation and `du -b` matched exactly.

Reply:
Fantastic!